docs/structured-streaming-kafka-integration.md

```diff
@@ -0,0 +1,231 @@
+      .readStream()
+      .format("kafka")
+      .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+      .option("subscribePattern", "topic.*")
+      .load()
+    ds3.selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+
+</div>
+<div data-lang="python" markdown="1">
+
+    # Subscribe to 1 topic
+    ds1 = spark
+      .readStream()
+      .format("kafka")
+      .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+      .option("subscribe", "topic1")
+      .load()
+    ds1.selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+
+    # Subscribe to multiple topics
+    ds2 = spark
+      .readStream
+      .format("kafka")
+      .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+      .option("subscribe", "topic1,topic2")
+      .load()
+    ds2.selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+
+    # Subscribe to a pattern
+    ds3 = spark
+      .readStream()
+      .format("kafka")
+      .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+      .option("subscribePattern", "topic.*")
+      .load()
+    ds3.selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+
+</div>
+</div>
+
+Each row in the source has the following schema:
+<table class="table">
+<tr><th>Column</th><th>Type</th></tr>
+<tr>
+  <td>key</td>
+  <td>binary</td>
+</tr>
+<tr>
+  <td>value</td>
+  <td>binary</td>
+</tr>
+<tr>
+  <td>topic</td>
+  <td>string</td>
+</tr>
+<tr>
+  <td>partition</td>
+  <td>int</td>
+</tr>
+<tr>
+  <td>offset</td>
+  <td>long</td>
+</tr>
+<tr>
+  <td>timestamp</td>
+  <td>long</td>
+</tr>
+<tr>
+  <td>timestampType</td>
+  <td>int</td>
+</tr>
+</table>
+
+Right now, the Kafka source has the following Spark's specific options.
+
+<table class="table">
+<tr><th>Option</th><th>value</th><th>default</th><th>meaning</th></tr>
+<tr>
+  <td>startingOffset</td>
+  <td>["earliest", "latest"]</td>
+  <td>"latest"</td>
+  <td>The start point when a query is started, either "earliest" which is from the earliest offset, 
+  or "latest" which is just from the latest offset. Note: This only applies when a new Streaming q
+  uery is started, and that resuming will always pick up from where the query left off.</td>
+</tr>
+<tr>
+  <td>failOnCorruptMetadata</td>
+  <td>[true, false]</td>
+  <td>true</td>
+  <td>Whether to fail the query when metadata is corrupt (e.g., topics are deleted, or offsets are 
+  out of range), which may lost data.</td>
+</tr>
+<tr>
+  <td>subscribe</td>
+  <td>A comma-separated list of topics</td>
+  <td>(none)</td>
+  <td>The topic list to subscribe. Only one of "subscribe" and "subscribePattern" options can be 
+  specified for Kafka source.</td>
+</tr>
+<tr>
+  <td>subscribePattern</td>
+  <td>Java regex string</td>
+  <td>(none)</td>
+  <td>The pattern used to subscribe the topic. Only one of "subscribe" and "subscribePattern" 
+  options can be specified for Kafka source.</td>
+</tr>
+<tr>
+  <td>kafka.consumer.poll.timeoutMs</td>
```

Comment:
> this is not a spark specific one. this can be in the must-be-specified table.

This is actually not a Kafka config. This is only used in KafkaSource